MM-29104: Add avatars component

ashishbhate opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A brief description of what this pull request does.
-->
Adds an avatar list component. It is required for the new collapsed reply threads feature. The feature design is at https://www.figma.com/file/t2KIwZnxb7h94249k6XdOp/MM-18107-Mobile-Threads-Overhaul and https://mattermost.atlassian.net/wiki/spaces/Threads/pages/1003945985/Mobile+User+Experience

The component is mostly done. The few things left to do
- Participants list screen (similar to the reactions list screen). This I will add in a separate PR.

The storybook story loads userIds from state, so before loading storybook try to visit as many user profiles as possible in the app. This will ensure that enough user profiles are loaded into the state. 

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
https://mattermost.atlassian.net/browse/MM-29104

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [ ] Includes text changes and localization file updates


#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) --> 
- iphone 11 simulator running ios 13.7
- Android simulator running Android 11

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs (for both iOS and Android if possible).
-->
Android storybook:

![Screenshot_1607326820](https://user-images.githubusercontent.com/1779129/101340873-4f3c2380-38a6-11eb-837c-a19449b8cbc0.png)
![Screenshot_1607326847](https://user-images.githubusercontent.com/1779129/101340880-52371400-38a6-11eb-977e-6195cdaf1335.png)

Iphone storybook:
![Simulator Screen Shot - iPhone 11 - 2020-12-07 at 13 18 36](https://user-images.githubusercontent.com/1779129/101340883-52cfaa80-38a6-11eb-90a4-d11e1d22dae4.png)
![Simulator Screen Shot - iPhone 11 - 2020-12-07 at 13 19 52](https://user-images.githubusercontent.com/1779129/101340885-53684100-38a6-11eb-82d5-d299e29d3303.png)

ashishbhate commented on app/components/avatars/avatars.stories.tsx:
```diff
@@ -0,0 +1,19 @@
+        <Avatars
+            size={number('size', 24, {range: true, min: 16, max: 80, step: 8})}
+            userIds={array('userIds', ['cb5wff5ah3y8ifijsgwrdhnn9o', 'psd3g9p9sjykuxqijh9qzmh86w', 'ue6q4zzgntykikm6f6fy5tju8r', 'iq164nrrmjff5mtbafr4jpukcy'], ',')}
+            theme={select('theme', Preferences.THEMES, Preferences.THEMES.default)}
```
Storybook complains about bad type here but still works. I think this is an issue with knobs

ashishbhate replied:
Changing the theme isn't very helpful since the Storybook theme does not change when this theme changes. Only the component theme changes, which looks a bit strange on the white storybook background.

ashishbhate commented on app/components/avatars/avatars.tsx:
```diff
@@ -0,0 +1,109 @@
+        const ss = getStyleSheet(theme, size);
+
+        return (
+            <TouchableWithFeedback>
```
The participants list press handler will go here. I'll be opening a separate PR for that screen soon.

ashishbhate commented on storybook/mattermost_storybook.ts:
```diff
@@ -2,7 +2,9 @@
 DevMenu.addItem('StoryBook', () => goToScreen('StoryBook', 'StoryBook'));
-Navigation.registerComponent('StoryBook', () => require('../storybook').default, () => require('../storybook').default);
+
+Navigation.registerComponent('StoryBook', () => withReduxProvider(require('../storybook').default), () => require('../storybook').default);
```
Hooking storybook up to our state context so that I can access user data (profile images in this case) using userIDs

enahum replied:
```suggestion
Navigation.registerComponent('StoryBook', () => withReduxProvider(require('../storybook').default));
```

this should work

ashishbhate commented on storybook/storyLoader.js:
```diff
@@ -1,13 +1,19 @@
-// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
-// See LICENSE.txt for license information.
+// Auto-generated file created by react-native-storybook-loader
+// Do not edit.
+//
+// https://github.com/elderfo/react-native-storybook-loader.git
```
I'm not sure why this changed, I think maybe our linter or pre-commit hook isn't working?

ashishbhate commented on app/components/avatars/avatars.stories.tsx:
```diff
@@ -0,0 +1,19 @@
+    add('Avatars', () => (
+        <Avatars
+            size={number('size', 24, {range: true, min: 16, max: 80, step: 8})}
+            userIds={array('userIds', ['cb5wff5ah3y8ifijsgwrdhnn9o', 'psd3g9p9sjykuxqijh9qzmh86w', 'ue6q4zzgntykikm6f6fy5tju8r', 'iq164nrrmjff5mtbafr4jpukcy'], ',')}
```
These userIds are specific to my test server. You can change these with userIds from your test server. UserIds are viewable in the "User Management" section of the system console.

enahum replied:
Please try and find a way not to hard code this ID's

ashishbhate replied:
@enahum 

I’ve been bashing my head against a wall for a week trying to figure this out. I’ve been trying to dynamically download the user data from the server.

1. Storybook react-native doesn’t support doing some async work like fetching of data via external API before rendering/adding a story. 
Storybook react (i.e. for the web) only just received support for doing this in version 6.0 (which was released on August). Storybook for react-native still does not support it. The way around this is to use a wrapper component, see this suggestion: https://github.com/storybookjs/storybook/issues/696#issuecomment-280368818

2. Promises seem to be broken in storybook for react-native.  https://github.com/storybookjs/react-native/issues/57
The fix for this is very hacky. This makes it impossible to use our configured client (`Client4`) to download any data from the server. 

At this point I’m not sure what to try next.

enahum replied:
but you added the ability to get data from the redux store, yes? so you should  be able to get user profiles that you already have stored there, right ?

ashishbhate replied:
Indeed, I was so fixated about getting the client to work that I forgot about the state. I've updated to use the state. I'll update the PR summary

ashishbhate commented:
I'm not sure why CI is failing on 

```

/home/circleci/mattermost-mobile/app/components/post_draft/quick_actions/quick_actions.test.js
  27:4  error  Newline required at end of file but not found  eol-last
```

I'm looking into it.

matthewbirtch commented:
@ashishbhate how do I review? Is there a link I can access?

ashishbhate commented:
>@ashishbhate how do I review? Is there a link I can access?

@matthewbirtch I don't know of a way to host this for you to review. I've written a storybook story for the component, but you'll still need to be able to run the app from this branch. Are you able to setup a dev environment for yourself (https://developers.mattermost.com/contribute/mobile/developer-setup/)? I can help you get set up. You don't need a real device, a simulator works too.

ashishbhate commented on app/components/avatars/avatars.tsx:
```diff
@@ -0,0 +1,109 @@
+import type {Theme} from '@mm-redux/types/preferences';
+import {changeOpacity, makeStyleSheetFromTheme} from '@utils/theme';
+
+const getStyleSheet = (appTheme: Theme, size: number) => {
+    const imgOverlap = -Math.floor(size * 0.4);
+    return makeStyleSheetFromTheme((theme: Theme) => {
+        return {
```
>Would this work?

I get a strange `Invariant Violation` error when I do this. 

This arrow function that is passed to `makeStyleSheetFromTheme` is only passed the `theme` when it is called in `makeStyleSheetFromTheme`. I use the extra closure to capture size as well 

I still have a golang hangover, so there may be a better way to do this.

matthewbirtch commented:
> > @ashishbhate how do I review? Is there a link I can access?
> 
> @matthewbirtch I don't know of a way to host this for you to review. I've written a storybook story for the component, but you'll still need to be able to run the app from this branch. Are you able to setup a dev environment for yourself (https://developers.mattermost.com/contribute/mobile/developer-setup/)? I can help you get set up. You don't need a real device, a simulator works too.

@ashishbhate I'm actually setup with a dev environment locally. I'll just need your help getting this specific branch up and running. I'll reach out over a DM in Mattermost and we'll go from there

matthewbirtch commented:
@ashishbhate a few things I've noticed:

1. The fallback icon is too close to the edges of the circles. Can we adjust to be more like the example image below (notice the padding)
<img width="48" alt="avatar-sample" src="https://user-images.githubusercontent.com/2040554/102543540-71d9f380-4081-11eb-8f5b-96314d63207a.png">

2. There seems to be something wrong with the white outline on the avatars - it's a jagged line - more noticeable at larger sizes (see screenshot below)

3. The avatars aren't quite aligned. The first avatar seems to be shifted (see screenshot below)
![image](https://user-images.githubusercontent.com/2040554/102543006-a305f400-4080-11eb-84c0-b8cf1598ec6f.png)

4. Similar to the above, the avatars seem to shift when it changes from 3 to 4 circles.

5. For the `+1`, can you double check to make sure it's `semibold` or `600` weight for the font? Can you also check the font size on this? It seems like it might be a touch smaller than what was designed for the 24x24 size.

ashishbhate commented:
@matthewbirtch 

I think you have the component scaled to an unrealistically large size. The avatar list component reuses the ProfilePicture component. The ProfilePicture is rendered at a size of 32 in other places in the app. The render padding, font size etc. don't scale linearly. We will need to tweak the styles for the component to look nice at extremely large sizes. 

I suggest reviewing the component set to size 32 (the default profile image size, the avatar list will likely be rendered even smaller). Currently the mobile components allow the size to be any number. The padding, radius, font size are calculated based on the size. This breaks down when the component is rendered super large. If we want to support super large sizes, then we should follow the webapp way of doing things - having a fixed set of sizes (extra small, small, medium, large, extra large), whose styles we can set specifically, but this will need to be part of a larger restyling effort or we might end up creating inconsistencies.

matthewbirtch commented:
> I think you have the component scaled to an unrealistically large size.

I realize this. However, I just used the large size to demonstrate the problem. The issues outlined above appear at all sizes from what I can tell.

ashishbhate commented:
>I realize this. However, I just used the large size to demonstrate the problem. The issues outlined above appear at all sizes from what I can tell.

The problem scales with size. The styling issues you have highlighted in your screenshot at not immediately apparent to me at lower sizes (of course, I do not have an eye for these things). Do you feel the imperfections in the screenshots in the OP (Size 40 - bigger than what we will use in the app) are blocker from a UI perspective? I understand things are not pixel perfect, but are they good enough for now (at the important sizes), to be polished later? 

I mentioned in my previous comment about how components on mobile can effectively be any size, and the styling attempts to do the best, within reasonable size constraints. If you think the styling issues are a blocking problem, could you: 

1. List out the numerical sizes at which you'd want this component to render (profile images render at 32)
2. Highlight the specific issues at all these specific sizes. I do not think the padding issue appears at size 32 and below, but I don't have an eye for these things - I appreciate you pointing out the exact issues.

matthewbirtch commented:
> The problem scales with size. The styling issues you have highlighted in your screenshot at not immediately apparent to me at lower sizes (of course, I do not have an eye for these things). Do you feel the imperfections in the screenshots in the OP (Size 40 - bigger than what we will use in the app) are blocker from a UI perspective? I understand things are not pixel perfect, but are they good enough for now (at the important sizes), to be polished later?

@ashishbhate I'm not sure I understand the problem. This is a pretty simple component from what I can tell. Am I missing something? An outline shouldn't render the way it's rendering (as identified above) so I'm worried about the construction of the UI elements. Even at the designed size (24x24) it doesn't seem like the outline was right. It doesn't provide enough of a separation from the avatar beneath it. Are you using a stroke on the circle shape or are you constructing this a different way?

I think I've made the two issues pretty clear with the screenshots provided, so I'm not sure how else to describe the problem.

If you need to get this approved to move forward, we can do that and come back to this, but I figured that this was such a simple component that getting it pixel perfect would be easily achieved.

ashishbhate commented:
>@ashishbhate I'm not sure I understand the problem. This is a pretty simple component from what I can tell. Am I missing something? An outline shouldn't render the way it's rendering (as identified above) so I'm worried about the construction of the UI elements. Even at the designed size (24x24) it doesn't seem like the outline was right. It doesn't provide enough of a separation from the avatar beneath it. Are you using a stroke on the circle shape or are you constructing this a different way?

@matthewbirtch It may appear simple from a UI perspective but I suspect the devil is in the details. If you'd like to get your hands dirty - the style is in the `makeStyleSheetFromTheme` function of the `avatars.tsx` file. It is very similar in format to regular CSS. Keep in mind that the styling is constrained by existing components and their behaviour. Feel free to DM me if something isn't clear. 

>I think I've made the two issues pretty clear with the screenshots provided, so I'm not sure how else to describe the problem.

I think you're missing a key point - the rendering logic changes with size. The screenshots you posted were for the an unrealistically large size. **There may be altogether different issues at smaller sizes** 

As above, it would be very helpful if you could list:

1. The exact sizes you want the component to render at (I've mentioned the size of other similar components in other comments). You've mentioned 24 in your comment. Is this the only size?
2. This bears repeating - **there may be altogether different issues at smaller sizes**. It would be great if you could scale the component to the sizes you decide from (1) and let me know if there are any **other** issues. 


>If you need to get this approved to move forward, we can do that and come back to this, 

Yes, lets get this approved to move forward and circle back to this later.

matthewbirtch commented:
> I think you're missing a key point - the rendering logic changes with size. The screenshots you posted were for the an unrealistically large size. **There may be altogether different issues at smaller sizes**

I'm not missing this point. the screenshots were indeed for the enlarged size, but just posted those so you could more easily see the issue. The issue appears on the smaller sizes as well.

24x24 is indeed the only size right now we are using this multi-avatar component for with the threads

I will approve this for now so you are not blocked, but if you could add a to-do to revisit these issues identified here that would be great.

matthewbirtch reviewed (approved):
Approving as noted, knowing we will revisit to sort out the identified issues here

ashishbhate commented:
@matthewbirtch I've updated the code to render the component only at size 24. It renders better now with minimal visual artefacts. I've also made an Jira ticket to track UI polish issues https://mattermost.atlassian.net/browse/MM-31765

@josephbaylon Not much to test here since it's still just an isolate component, but I've added some unit tests.

josephbaylon reviewed (approved):
LGTM. Per Ashish, unable to test the components yet but I smoke tested the app locally to see if I can build and install and login without issue at the least.
